addons/stock/models/product.py

```diff
@@ -244,7 +245,13 @@ def _search_ids(model, values):
             w_ids = set(Warehouse.browse(_search_ids('stock.warehouse', warehouse)).mapped('view_location_id').ids)
             if location:
                 l_ids = _search_ids('stock.location', location)
-                location_ids = w_ids & l_ids
+                # Keep all locations which are child of w_ids
+                parents = Location.browse(w_ids).mapped("parent_path")
+                location_ids = {
+                    loc.id
+                    for loc in Location.browse(l_ids)
+                    if any(loc.parent_path.startswith(parent) for parent in parents)
+                }
```

Comment:
A test has been added. Not sure if it is the right place. 
I also added an error in case no location was found, which would otherwise lead to a lookup in all locations